Смышляев Дмитрий

vafle228 opened the pull request:
@w1jtoo

vafle228 committed:
Создал новый проект

vafle228 committed:
Описал структуру классов для токенайзеров

vafle228 committed:
Добавил парсеры

vafle228 committed:
Добавил генератор html кода

vafle228 committed:
Реализация алгоритма
- Также переименовал токенайзер

vafle228 committed:
Merge pull request #1 from vafle228/architecture

Засетапил архитектуру

vafle228 committed:
Пока не в тдд (забыл)
- Сделал токенайзер
- Поправил ошибки в сканерах

vafle228 committed:
Довожу тесты 1.0
- Добавил тесты для сканера спец символов

vafle228 committed:
Довожу тесты 2.0
- Довез тесы для текстового сканера
- Поправил тесты для специального сканера

vafle228 committed:
Поправил токены
- Токены теперь не содержат строчку, но знают как ее найти
- Поправил сканеры, с учетом новой логики

vafle228 committed:
Добавил пробел как отдельный токен
- И поправил тесты до кучи

vafle228 committed:
Новый сканер
- Добавил сканер для чисел
- Добавил тесты для сканера чисел
- Отрефакторил другие сканеры

vafle228 committed:
Чуть поправил токены
- Теперь токен получает сорс текст, чтобы строить велью
- Также появился дебаг ТуСтринг метод

vafle228 committed:
Довожу тесты 3.0
- Довез тесты для токенизатора
- Добавил токенизатору новый сканер

vafle228 committed:
Поправил ААА концепцию в тесте
- Добавил ентер между актом и ассертом )

vafle228 committed:
Merge pull request #2 from vafle228/tokenizer

Готовый токенайзер

vafle228 committed:
Добавил матч тулы
- Написал тулзы для упрощения работы с матчами
- Накатил на это все тесты

vafle228 committed:
Простейший италик
- Добавил правило для примитивного италика

vafle228 committed:
Улучшил апи паттерн билдера
- Докинул ему добавление нескольких токен-типов
- Добавил повторение токен-типов

vafle228 committed:
Небольшой ренейминг
- Переименовал токен TEXT на токен WORD

vafle228 committed:
Ввел паттерн как отдельную сущность
- Теперь не лист тайпов, а настоящий паттерн

vafle228 committed:
Новые экстеншены
- Добавил звезду Клини
- Добавил первое совпадение

vafle228 committed:
Переосмысление метча
- Добавил метчу возможность билдить обратно лист

Seekard commented on cs/Markdown/Generator/HTMLGenerator.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Markdown.Generator;
+
+public class HTMLGenerator
```
Зачем себя ограничивать, может сразу IGenerator сделаем?)

vafle228 replied:
Согласен: сделаю, как начну генераторы пилить )

Seekard commented on cs/Markdown/Tokenizer/Tokens/Token.cs:
```diff
@@ -0,0 +1,20 @@
+﻿namespace Markdown.Tokenizer.Tokens;
+
+public class Token(TokenType tokenType, int begin, int length, string sourceText)
```
Это DataClass, в нем нет логики (совсем чуть-чуть :D), только данные.
Давай тогда заиспользуем [Record](https://metanit.com/sharp/tutorial/3.51.php)?

Seekard replied:
И для чего сюда передавать сразу весь текст `sourceText`?
Не говорю, что это плохо, просто мне не совсем понятно.
И еще на подумать, в aspnet есть такой концепт - Do not retrieve more data than is necessary. Он говорит про минимизацию запросов при http взаимодействии, но в т.ч. красиво перекладывается на код и доменные модели.

vafle228 replied:
Рекод сделаю :)

vafle228 replied:
```sourceText``` сюда передается с мыслью, чтобы токен мог достать свое значение из текста, в котором он существует. Соглашусь, что может выглядеть как переизбыток данных, но пока я еще не дописал полностью логику генератора хтмл, пусть он полежит тут (хотя бы для дебаг целей)

Seekard commented on cs/Markdown/Tokenizer/Tokens/Token.cs:
```diff
@@ -0,0 +1,20 @@
+    public int Length { get; } = length;
+    public TokenType TokenType { get; } = tokenType;
+
+    public string GetValue()
```
Мб свойством? Тут логики не особо много)

vafle228 replied:
Будет свойство, согл

Seekard commented on cs/Markdown/Tokenizer/Tokens/Token.cs:
```diff
@@ -0,0 +1,20 @@
+
+    public override string ToString()
+    {
+        return $"Token {TokenType} | Value \"{GetValue()}\"";
```
NIT: Лучше не переопределять `.ToString` без лишней потребности. Не совсем очевидное поведение. В плане, без углубления в код нельзя понять, что `.ToString` делает именно это (я бы первым делом подумал, что он возвращает `value`, это стандартное поведение таких методов)

p.s. а если хочешь оставить что-то подобное для дебага, то можно заюзать Externsion метод по типу `.ToDebugString`

vafle228 replied:
Да да, знаю, что переопределять ту стринг это харам. Но это осмысленный харам, так как я хотел, чтобы мне дебагер писал, что лежит в листе, когда я по шагам гуляю. А если делать как предложил ты, то среда не подтянет этот метод, когда я буду в дебаге смотреть код

Seekard commented on cs/Markdown/Tokenizer/Tokens/TokenType.cs:
```diff
@@ -0,0 +1,10 @@
+    UNDERSCORE,
+    SPACE,
+    NEW_LINE,
+    NUMBER,
```
А чем `NUMBER` отличается от `TEXT` с точки зрения обработки?

Seekard replied:
И похожий вопрос на засыпку: для чего хочешь использовать `SPACE`?)

vafle228 replied:
NUMBER и TEXT отличаются тем, что число с ```_``` воспринимается как число, а текст - как тег. SPACE же нужен, чтобы хендлить вариант, когда у меня будет разрыв тегов между слов ```wor_d hel_lo```. Если не определить пробел, то сканер вернет мне последовательность ```TEXT _ TEXT _ TEXT``` которую легко спутать с тегом

Seekard commented on cs/MarkdownTests/Tokenizer/Scanners/NumberScannerTest.cs:
```diff
@@ -0,0 +1,31 @@
+    [TestCase(" 123", 0)]
+    [TestCase("_\n ", 0)]
+    [TestCase("abcdifgh", 0)]
+    public void NumberScanner_Scan_ShouldScanNullFromText(string text, int begin)
```
NIT: можно поставить `begin` значение по умолчанию. Тогда не потеряется гибкость тестов, но код станет чуть чище :)

vafle228 replied:
Я с этим уже разобрался в своей "приватной" ветке ))

Seekard commented on cs/Markdown/Parser/Nodes/Node.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace Markdown.Parser.Nodes;
+
+public class Node
```
Допиши, что тут должно быть плиз, не совсем понимаю, чем от токена отличается(

vafle228 replied:
По существу, токен - язык автомата. Вместо того чтобы читать посимвольно, я объединяю группы каких-то символов в токены, даю им тайпы и уже работаю с ними. Ноды же это результаты работы рулов. По сути в нодах будет лежать непрерывная последовательность токенов, которая будет являться каким либо тегом. А древовидность нод позволит описывать вложение одного тега в другой

Seekard commented on cs/Markdown/Parser/Rules/ItalicRule.cs:
```diff
@@ -0,0 +1,12 @@
+
+public class ItalicRule : IParsingRule
+{
+    public Node Match(List<Token> tokens)
```
Из этого как-то должен получиться автомат 🤔
А точно достаточно входных параметром для этого?
В общем, сорян, не совсем понимаю, как оно будет работать, буду рад, если опишешь)

vafle228 replied:
Из этого действительно не получится автомат: параметров уже больше. На данный момент основная задача рулов - найти свой тег в листе токенов. Я начну описание с примитивных тегов, типо текста и италика. Далее, при описании тега выделения, я хочу переиспользовать ранее написанные теги, чтобы получить типо такого ```__ ITALIC | TEXT __```. Продолжая так делать, я хочу определить "корневой" тег, который будет запускать парсинг всего текста.

Говоря иначе, я хочу написать автомат, который строится на регвырах. Сами же регвыры я буду переиспользовать, чтобы написать еще более сложный регвыр. Финальный же автомат будет выглядить как or по всем получившимся регвырам.

Seekard commented on cs/Markdown/Program.cs:
```diff
@@ -0,0 +1,20 @@
+{
+    public static void Main(string[] args)
+    {
+        var markdown = "This _is_ a __sample__ markdown _file_.\n";
```
Советы категории Б: можешь вынести эту строчку в константу, чтобы потом, при добавлении новых примеров, можно было легко между ними переключаться :)

```
        const string firstExample = "This _is_ a __sample__ markdown _file_.\n";
        const string secondExample = "#This is another __sample__ markdown _file_";
        var markdown = firstExample;
````

vafle228 replied:
Окей, сделаю :)

Seekard commented on cs/Markdown/Tokenizer/MarkdownTokenizer.cs:
```diff
@@ -0,0 +1,26 @@
+            var token = scanners
+                .Select(sc => sc.Scan(markdown, begin))
+                .First(token => token is not null);
+            begin += token!.Length; tokenList.Add(token);
```
Перенеси на новую строчку плиз, глазкам больно ахаха

vafle228 replied:
Ахахахаха, ладно перенесу

Seekard commented on cs/Markdown/Tokenizer/Scanners/SpecScanner.cs:
```diff
@@ -0,0 +1,26 @@
+        if (tokenType is null) return null;
+        
+        var notNullType = (TokenType)tokenType;
+        return new Token(notNullType, begin, 1, markdown);
```
А как будешь разбираться с bold? Это, если что, `__`. Тут уже два символа. К GetTokenType тот же вопрос)

vafle228 replied:
Как я описывал выше: токены это язык, который мой автомат собирается распознавать. Поэтому болд для меня это +- такая вещь ```__ TEXT __```

Seekard replied:
Не. я к тому, что сейчас все заточено под 1 символ (абсолютно все). bold в 2 символа может добавить проблем, если сразу не подумать над ним.
Если в следующих коммитах уже есть решение, то гуд)

Seekard commented on cs/Markdown/Tokenizer/Scanners/SpecScanner.cs:
```diff
@@ -0,0 +1,26 @@
+    }
+    
+    public static bool CanScan(char symbol) 
+        => GetTokenType(symbol) != null;
```
А если у нас будет текст вида `Всем_привет!`?
Потом будешь разбирать в автомате? Кажется, все сломается. Ну, у меня ощущение такое.

vafle228 replied:
Оно не должно сломаться, так как автомат не найдет закрывающей ```_``` и просто проскочит на следующий символ

Seekard commented on cs/Markdown/Tokenizer/Scanners/SpecScanner.cs:
```diff
@@ -0,0 +1,26 @@
+
+    private static TokenType? GetTokenType(char symbol) => symbol switch
+    {
+        ' ' => TokenType.SPACE,
```
NIT: мб словарем сделать? Удобнее будет :)
https://marcduerst.com/2019/10/30/mapping-with-dictionary-instead-switch/

vafle228 replied:
Словарь так словарь: ща поправлю :)

Seekard replied:
Не, NIT - это буквально "докопаться". Такие треды можешь не исправлять, если не хочешь/считаешь свое решение лучше. Я тут просто варианты накидываю)

vafle228 committed:
Добавил правило для текста
- Немного пошаманил с неймингом
- Чутка переделал ноду

vafle228 committed:
Паттерн рул

vafle228 committed:
Обалдеть всего
- Отказался от метч объектов
- Поделил ноды на текст ноду и тег ноду
- Добавил разных экстеншенов, чтобы код был более читабельным

vafle228 committed:
Поддержка дабл андерскора

vafle228 committed:
Небольшая правка логики
- Теперь италик тег умеет определять, когда он метчит тег в слове или вне его

vafle228 committed:
Новые тесты
- Тесты на пробел перед закрывающим
- Тесты на пробел после открывающего

vafle228 committed:
Непарный тег тест
- Добавил тест для непарных тегов

vafle228 committed:
Merge pull request #3 from vafle228/parser

Parser

vafle228 committed:
Реорганизация кода

vafle228 committed:
Заготовки под жирный тег
- Добавил жирный тег
- Накинул булевы правила (на будущее)
- Написал простые тесты
- Перекинул пару файлов (снова)

vafle228 committed:
Болд тег
- Сделал красивый болд тег
- Написал на все тесты

Seekard commented on cs/Markdown/Parser/Nodes/Node.cs:
```diff
@@ -0,0 +1,7 @@
+﻿namespace Markdown.Parser.Nodes;
+
+public class Node(NodeType nodeType, int consumed)
```
Опять Data class)

Seekard replied:
Не понимаю, за что отвечает `int Consumed`. Это не Length случайно?

Seekard commented on cs/Markdown/Parser/Nodes/TextNode.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace Markdown.Parser.Nodes;
+
+public class TextNode(int start, int consumed, List<Token> source) : Node(NodeType.TEXT, consumed)
```
А вот тут точно нарушение SRP и всего вместе взятого (или я не понимаю).  Почему сюда весь source закидывается?.. Почему не только нужные токены?

Seekard commented on cs/Markdown/Parser/Rules/BoldRule.cs:
```diff
@@ -0,0 +1,12 @@
+{
+    public Node? Match(List<Token> tokens, int begin = 0)
+    {
+        throw new NotImplementedException();
```
А хде...

Seekard commented on cs/Markdown/Parser/Rules/ItalicRule.cs:
```diff
@@ -0,0 +1,46 @@
+    }
+    
+    private static TagNode BuildNode(TextNode textNode) 
+        => new(NodeType.ITALIC, textNode, textNode.Consumed + 2);
```
Я бы не завязывался на `+ 2`. В md есть разные форматы, поэтому лучше вынести в константу

Seekard replied:
И как будто есть смысл вынести такой Factory метод в сами ноды, как думаешь?

Seekard commented on cs/Markdown/Parser/Rules/PatternRule.cs:
```diff
@@ -0,0 +1,22 @@
+    public Node? Match(List<Token> tokens, int begin = 0)
+    {
+        if (pattern.Count == 0) return null;
+        if (tokens.Count - begin < pattern.Count) return null;
```
NIT: почему думаешь, что лучше вернуть null, чем выкинуть Exception?

Seekard commented on cs/Markdown/Parser/Rules/Tools/ListMatchExtensions.cs:
```diff
@@ -0,0 +1,39 @@
+        {
+            var node = pattern.Match(tokens, begin);
+            if (node is null) return nodes;
+            begin += node.Consumed; nodes.Add(node);
```
перенос

Seekard commented on cs/Markdown/Tokenizer/Tokens/ListExtensions.cs:
```diff
@@ -0,0 +1,9 @@
+public static class ListExtensions
+{
+    public static string ToText(this List<Token> tokens) 
+        => tokens.Aggregate(new StringBuilder(), (sb, t) => sb.Append(t.Value)).ToString();
```
`.Join`?

Seekard commented on cs/Markdown/Parser/Rules/TextRule.cs:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    private static bool IsText(Token token)
+        => token.TokenType is TokenType.WORD or TokenType.SPACE;
```
Number? New Line?

vafle228 committed:
Фикс италик тега
- Италик тег стал еще более устойчив
- Код похорошел, новые плюхи появились, плохие ушли

vafle228 committed:
Уточнил правило
- Более семантично теперь

vafle228 committed:
Уточнил болд тег
- Более стойкий болд тег
- Более строгие тесты

vafle228 committed:
Подготовка к новыйм тегам
- Добавил новый токен
- Добавил новый тип нод
- Поправил тестики

vafle228 committed:
Тег параграфа
- Добавил тег параграфа
- Накидал тестов для тега
- Поправил тег инворд италик тег
- Поправил инворд болд тег

vafle228 committed:
Движение к будущему
- Начал рефакторинг тегов

vafle228 committed:
Новый бордер тег

vafle228 committed:
Нафиг эту бордер рулу вообщем
- Вернулся к кондитионал руле
- Честно прописал условия для границ тегов подчерка

vafle228 committed:
Убрал тупую нал проверку

vafle228 committed:
Отрефакторил параграф рул

vafle228 committed:
Отрефакторил хедлайн

vafle228 committed:
Поделил ответсвенность тестов
- Некоторые тесты перекочевали в InWord аналоги

vafle228 committed:
Обалдеть изменений
- Решил переименовать булевы рулы в специальный рулы (чем они и являются)

vafle228 committed:
Тесты для специальных рулзов
- Теперь все еквивалент матчи строги к порядку

vafle228 committed:
Ор рул тесты

vafle228 committed:
Тесты для кондитионал рулы

vafle228 committed:
Крупное переосмысление
- Теперь ноды не хранят в себе лист токенов
- Пошел править все ноды с этим обновлением

vafle228 committed:
Переделка италик тега
- Переделал италик тег на новые ноды
- Поменял токены для уменьшения их ответственности

vafle228 committed:
Переписал болд тег на новый лад

vafle228 committed:
Дофикс оставшихся тегов

vafle228 committed:
Избавился от лишних экстеншенов

vafle228 committed:
Тесты для звезды клини

vafle228 committed:
Тесты для континюс рулы

vafle228 committed:
Тесты для заголовка
- Добавил тесты для заголовка
- Переименовал метчи в ноды в некоторых тестках

vafle228 committed:
Ескейп тег
- Добавил новый тип токена
- Поправил тесты и сканеры
- Поправил рулы параграфа
- Написал сам тег + тесты

vafle228 committed:
Узнал про крутые строки

vafle228 committed:
Правки по ескейп тегу
- Ескейп тег теперь настоящий тег
- Ескейп тег принимает токены, которые он будет ескейпить
- Параграф теперь умеет ескейпить все андерскоры

vafle228 committed:
Бади рул
- Добавил тестики для бади рулы
- Написал саму бади рулу

vafle228 committed:
Merge pull request #4 from vafle228/parser

Parser

vafle228 committed:
Переместил тулзы на уровень выше
- Вроде тулзы переместил, а изменений обалдеть

vafle228 committed:
Добавил интерфейс генератора

vafle228 committed:
Реализовал хтмл генератор
- Оч строгий генератор на рекурсивном обходе дерева

vafle228 committed:
Оч строгие тесты

vafle228 committed:
Тесты на генератор хтмл

vafle228 committed:
Тест на скорость работы

vafle228 committed:
Починил бади тесты
- Из-за нового ньюлайн токена, тесты попадали (увы)

vafle228 committed:
Переписал токены на мемори спаны
- Теперь оно оч быстрое (линейное)

vafle228 committed:
Добавил уровни для хедлайна
- Появилась хедлайн нода
- Хедлайн тег научился строить хедлайн ноду
- Вернулся строгий метод Third()

vafle228 committed:
Уточнил болд и италик рулу
- Теперь внутри болд и италик рулы можно писать теговые символы

vafle228 committed:
Merge pull request #5 from vafle228/generator

Generator

vafle228 committed:
Хочу открыть мр
- Пустые мр нельзя открыть (звой)

vafle228 committed:
Добавил новые токены
- Добавил токены квадратных скобок
- Добавил токены обычных скобок

vafle228 committed:
Добавил href рулзу

vafle228 committed:
Описал спецификацию

vafle228 committed:
Внедрил хрефу в параграф

vafle228 committed:
Добавил хрефу в генератор хтмл

vafle228 committed:
Перенес спеку в маркадаун файл

vafle228 committed:
Чуть прибрался

vafle228 committed:
Merge pull request #6 from vafle228/custom-tag

Custom tag
